rclpy/src/rclpy/_rclpy.c

```diff
@@ -3369,6 +3370,213 @@ rclpy_clock_set_ros_time_override(PyObject * Py_UNUSED(self), PyObject * args)
+    }
+  }
+  return true;
+}
+
+/// Create an rclpy.parameter.Parameter from an rcl_variant_t
+/**
+ * On failure a Python exception is raised and false is returned if:
+ *
+ * Raises ValueError if the variant points to no data.
+ *
+ * \param[in] name The name of the parameter
+ * \param[in] variant The variant object to create a Parameter from
+ * \param[in] parameter_cls The PythonObject for the Parameter class.
+ * \param[in] parameter_cls The PythonObject for the Parameter.Type class.
+ *
+ * Returns a pointer to an rclpy.parameter.Parameter with the name, type, and value from
+ * the variant or NULL when raising a python exception.
+ */
+static PyObject * _parameter_from_rcl_variant(
+  const char * name, rcl_variant_t * variant, PyObject * parameter_cls,
+  PyObject * parameter_type_cls)
+{
+  int type_enum_value;
+  PyObject * value;
+  if (variant->bool_value) {
+    type_enum_value = 1;
+    value = variant->bool_value ? Py_True : Py_False;
+  } else if (variant->integer_value) {
+    type_enum_value = 2;
+    value = PyLong_FromLong(*(variant->integer_value));
+  } else if (variant->double_value) {
+    type_enum_value = 3;
+    value = PyFloat_FromDouble(*(variant->double_value));
+  } else if (variant->string_value) {
+    type_enum_value = 4;
+    value = PyUnicode_FromString(variant->string_value);
+  } else if (variant->byte_array_value) {
+    type_enum_value = 5;
+    value = PyList_New(variant->byte_array_value->size);
+    for (size_t i = 0; i < variant->byte_array_value->size; i++) {
+      PyList_SetItem(value, i, PyBytes_FromFormat("%u", variant->byte_array_value->values[i]));
+    }
+  } else if (variant->bool_array_value) {
+    type_enum_value = 6;
+    value = PyList_New(variant->bool_array_value->size);
+    for (size_t i = 0; i < variant->bool_array_value->size; i++) {
+      PyList_SetItem(value, i, variant->bool_array_value->values[i] ? Py_True : Py_False);
+    }
+  } else if (variant->integer_array_value) {
+    type_enum_value = 7;
+    value = PyList_New(variant->integer_array_value->size);
+    for (size_t i = 0; i < variant->integer_array_value->size; i++) {
+      PyList_SetItem(value, i, PyLong_FromLong(variant->integer_array_value->values[i]));
+    }
+  } else if (variant->double_array_value) {
+    type_enum_value = 8;
+    value = PyList_New(variant->double_array_value->size);
+    for (size_t i = 0; i < variant->double_array_value->size; i++) {
+      PyList_SetItem(value, i, PyFloat_FromDouble(variant->double_array_value->values[i]));
+    }
+  } else if (variant->string_array_value) {
+    type_enum_value = 9;
+    value = PyList_New(variant->string_array_value->size);
+    for (size_t i = 0; i < variant->string_array_value->size; i++) {
+      PyList_SetItem(value, i, PyUnicode_FromString(variant->string_array_value->data[i]));
+    }
+  }
+
+  PyObject * type = PyObject_CallObject(parameter_type_cls, Py_BuildValue("(i)", type_enum_value));
```

Comment:
Need to check `Py_BuildValue` didn't return `NULL`, as well as `PyObject_CallObject`.